feat: add TypeScript type annotation tests with node 22 snapshots

mertcanaltin opened the pull request:
Greetings, I added a test and tried to get more coverage

- Add test for stripping type annotations from functions
- Add test for stripping type annotations from classes
- Add test for stripping type annotations from interfaces
- Add test for stripping type annotations from type aliases
- Add test for stripping type annotations from generics
- Add test for stripping type annotations from arrow functions
- Add test for stripping type annotations from type assertions

Additionally, I have integrated node 22 snapshot testing as suggested. The snapshot tests have been normalized to handle line endings and formatting issues

marco-ippolito commented:
Maybe we can use node 22 snapshots testing?

mertcanaltin commented:
> Maybe we can use node 22 snapshots testing?

do you think I should integrate it here?

marco-ippolito commented:
> > Maybe we can use node 22 snapshots testing?
> 
> do you think I should integrate it here?

Id give it a try, and we can use node 22 in the CI

mertcanaltin commented:
Hi, I converted it to a snapshot

marco-ippolito commented:
missing EOF and some formatting issues

marco-ippolito commented on test/snapshot.js:
```diff
@@ -1,33 +1,38 @@
-const fs = require("node:fs");
-const path = require("node:path");
+const fs = require("fs");
```
please use `node:` namespace for imports

mertcanaltin replied:
ı used, thanks

marco-ippolito commented on package.json:
```diff
@@ -23,7 +23,7 @@
 		"build": "rspack build",
 		"typecheck": "tsc --noEmit",
-		"test": "node --test ./test"
+		"test": "node --test --experimental-test-snapshots --test-update-snapshots --require ./test/snapshot-config.js ./test/index.test.js"
```
the creation of snapshots should be run in a different command like
`test:regenerate`

mertcanaltin commented:
I updated the test structure, it seemed that there was no need for the main index test anymore, so I updated the test commands in the package.json, but if any tests are added, we can update here

AugustinMauroy commented:
I think we can add test for this code:
```ts
type User = {
  name: string;
  age: number;
};

function isAdult(user: User): boolean {
  return user.age >= 18;
}

const justine = {
  name: 'Justine',
  age: 23,
} satisfies User;

const isJustineAnAdult = isAdult(justine);
```

it will be added to the learn section as an example it would be interesting to test it

AugustinMauroy commented on test/setup.mjs:
```diff
@@ -0,0 +1,4 @@
+import { register } from "node:module";
```
this file isn't use we can remove it ?

mertcanaltin replied:
Of course, thank you

marco-ippolito commented on test/snapshots/index.test.js:
```diff
@@ -0,0 +1,80 @@
+const { test } = require("node:test");
```
shouldnt this test be in the root ? also why remove white space? the whole point is to check for whitespaces

mertcanaltin replied:
I thought we were gonna keep the snapshots tests separate

marco-ippolito replied:
The test should stay in the test folder, snapshot output in the test/snapshot folder

mertcanaltin replied:
I tried to fix it

mertcanaltin commented:
> I think we can add test for this code:
> 
> ```ts
> type User = {
>   name: string;
>   age: number;
> };
> 
> function isAdult(user: User): boolean {
>   return user.age >= 18;
> }
> 
> const justine = {
>   name: 'Justine',
>   age: 23,
> } satisfies User;
> 
> const isJustineAnAdult = isAdult(justine);
> ```
> 
> it will be added to the learn section as an example it would be interesting to test it

I have added a test on this, thanks for the suggestion

marco-ippolito commented on package.json:
```diff
@@ -23,7 +23,8 @@
 		"build": "rspack build",
 		"typecheck": "tsc --noEmit",
-		"test": "node --test ./test"
+		"test": "node --test --experimental-test-snapshots ./test/index.test.js",
```
We want to execute the whole folder rathee then a single file

marco-ippolito commented on package.json:
```diff
@@ -23,7 +23,8 @@
 		"typecheck": "tsc --noEmit",
-		"test": "node --test ./test"
+		"test": "node --test --experimental-test-snapshots ./test/index.test.js",
+		"test:regenerate": "node --experimental-test-snapshots --test-update-snapshots ./test/index.test.js"
```
Same here

mertcanaltin replied:
```js
"test": "node --test ./test"
```

ı used this commant, but get a this error in tests

```js
ℹ tests 1
ℹ suites 0
ℹ pass 0
ℹ fail 1
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 86.965917

✖ failing tests:

test at test:1:1
✖ test (79.325333ms)
  'test failed'
➜  amaro git:(dev-1) ✗ 
```

marco-ippolito replied:
Thats odd try to remove the path completely

mertcanaltin replied:
```js
"test": "node --test --experimental-test-snapshots",
"test:regenerate": "node --test --experimental-test-snapshots --test-update-snapshots ./test"
```

unfortunately I failed I updated this way

mertcanaltin replied:
this problem was solved in your last suggestion thank you

marco-ippolito commented:
this should work
```
		"test": "node --test --experimental-test-snapshots ./test/*.test.js",
		"test:regenerate": "node --test --experimental-test-snapshots --test-update-snapshots ./test/*.test.js"
```

marco-ippolito commented:
Ci needs to be run with node 22 and also the engine field with node => 22 in the package.json

mertcanaltin commented:
yes it worked as you suggested I tried at 22 🚀

marco-ippolito commented:
please update the ci.yml to use node 22 and add the engines in the package json

marco-ippolito commented on .github/workflows/ci.yml:
```diff
@@ -10,19 +10,8 @@ env:
   YARN_ENABLE_GLOBAL_CACHE: false
 
 jobs:
-  commit-lint:
```
Why remove commit lint, you need to rebase

mertcanaltin replied:
I forgot to update the main branch, so sorry, this linti crushed it 😢

mertcanaltin replied:
ı solved

marco-ippolito commented on test/setup.mjs:
```diff
@@ -0,0 +1,3 @@
+register("some-typescript-loader");
```
Whats the point of this?

mertcanaltin replied:
I thought it was necessary because it was in the document, but we don't need it, I think I take it back
